magma/backend/mlir/hardware_module.py

```diff
@@ -373,57 +421,91 @@ def visit_magma_mux(self, module: ModuleWrapper) -> bool:
         # magma/primitives/mux.py.
         height = len(list(filter(
             lambda p: "I" in p.name.name, defn.interface.outputs())))
-        T = type(defn.I0)
-        mlir_type = hw.ArrayType((height,), magma_type_to_mlir_type(T))
-        array = self._ctx.new_value(mlir_type)
-        hw.ArrayCreateOp(
-            operands=module.operands[:-1],
-            results=[array])
-        hw.ArrayGetOp(
-            operands=[array, module.operands[-1]],
-            results=module.results)
+        data, select = module.operands[:-1], module.operands[-1]
+        # NOTE(rsetaluri): This is a specialized code path for the case where
+        # all tuples are flattened.
+        if self._ctx.opts.flatten_all_tuples and len(data) != height:
+            assert len(data) % height == 0
+            stride = len(data) // height
+            assert len(module.results) == stride
+            for i in range(stride):
+                operands = data[i::stride]
+                make_mux(data[i::stride], select, module.results[i])
+            return True
+        make_mux(data, select, module.results[0])
         return True
 
     @wrap_with_not_implemented_error
     def visit_magma_register(self, module: ModuleWrapper) -> bool:
+        # NOTE(rsetaluri): Refactoring this compilation into the make_register
+        # and get_data_and_init functions is necessary to support
+        # flatten_all_tuples=True. make_register works for arbitrary types of
+        # inputs, so we can either invoke it once for the top-level type or call
+        # it on all flattened parts. get_data_and_init associates the operands
+        # with init values (and magma type), so that flattened and unflattened
+        # tuples can be treated uniformly.
+
+        def make_register(T, data, init, result):
+            reg = self._ctx.new_value(hw.InOutType(data.type))
+            sv.RegOp(name=inst.name, results=[reg])
+            always = sv.AlwaysFFOp(operands=always_operands, **attrs)
+            const = self.make_constant(T, init)
+            with push_block(always.body_block):
+                ctx = contextlib.nullcontext()
+                if has_enable:
+                    ctx = push_block(sv.IfOp(operands=[enable]).then_block)
+                with ctx:
+                    sv.PAssignOp(operands=[reg, data])
+            if has_reset:
+                with push_block(always.reset_block):
+                    sv.PAssignOp(operands=[reg, const])
+            with push_block(sv.InitialOp()):
+                sv.BPAssignOp(operands=[reg, const])
+            sv.ReadInOutOp(operands=[reg], results=[result])
```

Comment:
Similar, should we move this to be an internal method?

Reply:
this definitely looks wonky, but needs to be a closure. many of the common args such as always_operands, enable, has_reset are shared across the invokations of this function. ideally would be all factored out, into functions with good interfaces -- this is hacky but is mostly clean